### PR TITLE
Fix #2808, remove endorsement field CRL_ISSUER_CHAIN_PCK_PROC_CA.

### LINKED ISSUE #2808: Duplicate info in sgx endorsements data structure
In current design of oe_sgx_endorsements_fields (shown below, also can be found in https://github.com/openenclave/openenclave/tree/master/docs/DesignDocs), the field OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA is not necessary and is not being used. 

The OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT already contains the info in OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA

typedef enum _oe_sgx_endorsements_fields
{
    OE_SGX_ENDORSEMENT_FIELD_VERSION,
    OE_SGX_ENDORSEMENT_FIELD_TCB_INFO,
    OE_SGX_ENDORSEMENT_FIELD_TCB_ISSUER_CHAIN,
    OE_SGX_ENDORSEMENT_FIELD_CRL_PCK_CERT,
    OE_SGX_ENDORSEMENT_FIELD_CRL_PCK_PROC_CA,
    **OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT,
    OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA,**
    OE_SGX_ENDORSEMENT_FIELD_QE_ID_INFO,
    OE_SGX_ENDORSEMENT_FIELD_QE_ID_ISSUER_CHAIN,
    OE_SGX_ENDORSEMENT_FIELD_CREATION_DATETIME,
    OE_SGX_ENDORSEMENT_COUNT
} oe_sgx_endorsements_fields;

The CRL issuer chains info is being validated in oe_validate_revocation_list. https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/collateral.c#L214
The CRL issuer chain array  was defined at the beginning of the function.
The first element of the array is pck cert CRL issuer chain, and the second one is the processor CA CRL issuer chain.

https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/collateral.c#L224

Then it is passed as a parameter to function oe_cert_verify and the only the first element (OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT) in that array was verified (Only verifying OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT actually is enough and correct behavior here). 
https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/collateral.c#L320

We may need to remove the OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA from the data structure and do other related changes.

### PR DESCRIPTION
**Pull request by qiucwang**
Fixes #2808 
Removed OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA from endorsement struct.

Current endorsement struct looks like:
typedef enum _oe_sgx_endorsements_fields_t
{
    OE_SGX_ENDORSEMENT_FIELD_VERSION,
    OE_SGX_ENDORSEMENT_FIELD_TCB_INFO,
    OE_SGX_ENDORSEMENT_FIELD_TCB_ISSUER_CHAIN,
    OE_SGX_ENDORSEMENT_FIELD_CRL_PCK_CERT,
    OE_SGX_ENDORSEMENT_FIELD_CRL_PCK_PROC_CA,
    **OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT,
    OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA,**
    OE_SGX_ENDORSEMENT_FIELD_QE_ID_INFO,
    OE_SGX_ENDORSEMENT_FIELD_QE_ID_ISSUER_CHAIN,
    OE_SGX_ENDORSEMENT_FIELD_CREATION_DATETIME,
    OE_SGX_ENDORSEMENT_COUNT
} oe_sgx_endorsements_fields_t;

In `oe_create_sgx_endorsements()`, both fields are copied from `quote_verification_collateral->pck_crl_issuer_chain` so they have the identical content.
https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/endorsements.c#L65-L69
First `buffer` points to `ISSUER_CHAIN_PCK_CERT` and second one points to `ISSUER_CHAIN_PCK_PROC_CA`
https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/endorsements.c#L196-L212
And later in `oe_validate_revocation_list()`, both fields are wrapped into a `oe_cert_chain_t crl_issuer_chain[3]` 
https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/collateral.c#L224
https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/common/sgx/collateral.c#L283-L293
and used to verify the `PCK_CERT` by `oe_cert_verify()`. But in `oe_cert_verify()`, crl_issuer_chain is treated as `oe_cert_chain_t*` and only the first element, `ISSUER_CHAIN_PCK_CERT`,  is used to verify the `cert`. And one issuer chain is enough to verify one certificate chain.
https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/include/openenclave/internal/crypto/cert.h#L126-L146
So `ISSUER_CHAIN_PCK_PROC_CA` is duplicated with `ISSUER_CHAIN_PCK_CERT` and also not used in verification. So we could remove it.

Signed-off-by: Qiucheng Wang <qiucwang@microsoft.com>

### CODE REVIEW

File: docs/DesignDocs/RemoteAttestationCollaterals.md [hunk 1]
```diff
@@ -327,11 +326,9 @@ typedef struct _oe_sgx_endorsements_t
      *      crl[0] = CRL for the SGX PCK Certificate
      *      crl[1] = CRL for the SGX PCK Processor CA
      *
-     *  OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT to
-     *     OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA
+     *  OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT
      *  CRLs issuer chains in PEM format, null-terminated string
-     *      crl[0] = Issuer Chain for the SGX PCK Certificate
-     *      crl[1] = CRL for the SGX PCK Processor CA
+     *      Issuer Chain for the SGX PCK Certificate
```
Comment (yufjiang) [inline docs/DesignDocs/RemoteAttestationCollaterals.md:331]: The Issuer Chain for the SGX PCK Certificate **CRL**. Actually this info has been described by the sentence above, you might just remove it.

File: docs/DesignDocs/RemoteAttestationCollaterals.md [hunk 2]
```diff
@@ -327,11 +326,9 @@ typedef struct _oe_sgx_endorsements_t
      *      crl[0] = CRL for the SGX PCK Certificate
      *      crl[1] = CRL for the SGX PCK Processor CA
      *
-     *  OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT to
-     *     OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA
+     *  OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT
      *  CRLs issuer chains in PEM format, null-terminated string
```
Comment (yufjiang) [inline docs/DesignDocs/RemoteAttestationCollaterals.md:330]: Maybe rewording to "The SGX PCK CRL issuer chain in PEM format," we only have one issuer chain now.

**Comment by yufjiang** [comment]
bors try

File: include/openenclave/bits/attestation.h [hunk 3]
```diff
@@ -85,7 +85,6 @@ typedef enum _oe_sgx_endorsements_fields_t
     OE_SGX_ENDORSEMENT_FIELD_CRL_PCK_CERT,
     OE_SGX_ENDORSEMENT_FIELD_CRL_PCK_PROC_CA,
     OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_CERT,
-    OE_SGX_ENDORSEMENT_FIELD_CRL_ISSUER_CHAIN_PCK_PROC_CA,
```
Comment (jazzybluesea) [inline include/openenclave/bits/attestation.h:88]: Since this is a public header, I suggest adding a note in the release notes for this change.

Comment (qiucwang) [reply include/openenclave/bits/attestation.h:88]: Thanks. I checked with @radhikaj  and she feels it's fine to add that in the CHANGE.log as I'm fixing a bug.

File: docs/DesignDocs/RemoteAttestationCollaterals.md (hunk 1 above)
Comment (qiucwang) [reply docs/DesignDocs/RemoteAttestationCollaterals.md]: Thanks. Updated that.

**Comment by jazzybluesea** [comment]
@qiucwang As discussed, please make sure that we no longer need this field now or in the future.  Perhaps there is some verification that needs to be done that we are not doing yet?

**Comment by qiucwang** [comment]
> @qiucwang As discussed, please make sure that we no longer need this field now or in the future. Perhaps there is some verification that needs to be done that we are not doing yet?

Sure. And now we are using `crl_issuer_chain` to verify against certificate chain, which is done in `oe_cert_verify()`. That's the verification related with the issuer_chain. And there is only one issuer_chain from one collateral, which is `quote_verification_collateral->pck_crl_issuer_chain`. We duplicated it and saved in two fields in endorsement so one crl_issuer_chain field matches one crl field. But verifying crl against certificate and issuer_chain against certificate is enough. And `oe_cert_verify()` just takes one chain as parameter as well. So we don't need a duplicated copy.

**Comment by yufjiang** [comment]
bors try

File: include/openenclave/bits/attestation.h (hunk 3 above)
Comment (yentsanglee) [reply include/openenclave/bits/attestation.h:88]: The API that uses these fields are in endorsements.h which is not public. So we are ok.

---
In reply to: [424736127](https://github.com/openenclave/openenclave/pull/3018#discussion_r424736127) [](ancestors = 424736127)

File: common/sgx/collateral.c [hunk 4]
```diff
@@ -247,9 +247,6 @@ oe_result_t oe_validate_revocation_list(
             version,
             OE_SGX_ENDORSEMENTS_VERSION);
 
-    OE_STATIC_ASSERT(
-        OE_COUNTOF(crl_issuer_chain) >= OE_SGX_ENDORSEMENTS_CRL_COUNT);
```
Comment (yentsanglee) [inline common/sgx/collateral.c:251]: This assertion should stay and check 'crls' instead because OE_SGX_ENDORSEMENTS_CRL_COUNT is the max index for writing to crls[i].

Comment (qiucwang) [reply common/sgx/collateral.c:251]: Updated.

**Comment by yentsanglee** [comment]
bors +

**Comment by qiucwang** [comment]
bors try

File: include/openenclave/bits/attestation.h (hunk 3 above)
Comment (dthaler) [reply include/openenclave/bits/attestation.h:88]: At some point, we should move SGX-specific stuff out of this non-SGX-specific header, but not in this PR.
